### DISCUSSION TITLE
neo-tree always opens CWD instead of root ( Yes I tried both e and E. it opens CWD on both occasions)

### DISCUSSION THREAD

**Discussion by ash77ash77** [LazyVim/LazyVim#4768] [Q&A] [answered]
And no I didn't touch the config files at all. I simply installed a couple of languages from LazyExtra and don't configure anything there either. it used to work just fine before. I think a recent update broke it. And tnx in advance for help. much appreciated

**Comment by dpetka2001** [comment]
The root that you mention is the `root_dir` that `LazyVim.root` resolves based on either LSP `root_dir`, pattern or current working directory. When you open Neovim at the root of a LSP workspace, then the `root_dir` the corresponding LSP server reports back and the `cwd` coincide. It has been like that for a really long time.

**Reply by ash77ash77** [reply]
Thanks for replying much appreciated. But unfortunately I can't grasp exactly what you're conveying here fully. Is what you're saying is because of my LSP, the root becomes the place where I opened neovim in? As in my current project directory? And to add to this, it's not just neo-tree, fzf and grep also does the exact same thing. Like I said before, it wasn't like this at all. Neo-tree , fzf, live grep all used to open by default in root just fine unless specifically told to open in cwd. I had the exact same plugins when it used to work fine before.

**Reply by dpetka2001** [reply]
It's been like this for a really long time. Just do `:LspInfo` and you'll see the `root_dir` that gets returned by the LSP server. There are chances that it will be the same as the current directory if you open Neovim in the same directory.

**Reply by dennypenta** [reply]
@dpetka2001 do you mind to add a little more on Lazyvim.root? 
I read the source and can't understand, I see it's used in neotree plugin.
Im asking to understand how to work with that and if I should use it.

For instance, if I open a project and then navigate to a file that is in another module (my dependency package) and input `vim.uv.cwd` and `LazyVim.root()` they give me exact same result pointing to the project directory I opened.

**Reply by dpetka2001** [reply, marked as answer]
`LazyVim.root` returns the `root_dir` based on 3 factors:
- If a LSP server is attached then it will return the `root_dir` that is detected by the LSP server
- Then it searches for `.git` or `lua` patterns to determine the `root_dir` if a LSP server is not attached or doesn't detect a `root_dir` based on the root markers that each LSP server expects to find to determine the `root_dir`
- As a last result it falls back to `cwd`, which is the directory where Neovim was opened from

Usually the lower-case letters use `LazyVim.root` and the capital-case letters use directly `cwd` (for example `<leader>e` opens in detected `root_dir` based on the 3 factors I mentioned above, whereas `<leader>E` opens directly in `cwd` which is the directory where Neovim was opened from).

If you type the command `:LazyRoot` it will show you the `root_dir` that is detected and which of the 3 factors I described above, was the one responsible for the determination of `root_dir`.

**Reply by dpetka2001** [reply]
You can change the behavior/priority of the 3 factors i mentioned by defining `vim.g.root_spec` in your personal configuration in `~/.config/nvim/lua/config/options.lua` or changing directly the keymaps according to your personal preferences.
